src/main/java/com/linecorp/armeria/client/metrics/MetricCollectingClientCodec.java

```diff
@@ -0,0 +1,102 @@
+package com.linecorp.armeria.client.metrics;
+
+import java.lang.reflect.Method;
+import java.util.Optional;
+import java.util.function.LongSupplier;
+
+import com.linecorp.armeria.client.ClientCodec;
+import com.linecorp.armeria.client.DecoratingClientCodec;
+import com.linecorp.armeria.common.Scheme;
+import com.linecorp.armeria.common.SerializationFormat;
+import com.linecorp.armeria.common.ServiceInvocationContext;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.metrics.MetricConsumer;
+import com.linecorp.armeria.common.metrics.MetricsData;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufHolder;
+import io.netty.channel.Channel;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.util.AttributeKey;
+
+/**
+ * Decorator to collect client metrics.
+ * <p>
+ * This class is expected to be used with other {@link MetricConsumer}
+ */
+class MetricCollectingClientCodec extends DecoratingClientCodec {
+
+    private static final AttributeKey<MetricsData> METRICS =
+            AttributeKey.valueOf(MetricCollectingClientCodec.class, "METRICS");
+
+    private final MetricConsumer metricConsumer;
+
+    /**
+     * Creates a new instance that decorates the specified {@link ClientCodec}.
+     */
+    MetricCollectingClientCodec(ClientCodec delegate, MetricConsumer metricConsumer) {
+        super(delegate);
+        this.metricConsumer = metricConsumer;
+    }
+
+    @Override
+    public EncodeResult encodeRequest(Channel channel, SessionProtocol sessionProtocol, Method method,
+                                      Object[] args) {
+        long startTime = System.nanoTime();
+        EncodeResult result = delegate().encodeRequest(channel, sessionProtocol, method, args);
+        LongSupplier lazyElapsedTime = () -> System.nanoTime() - startTime;
+        if (result.isSuccess()) {
+            ServiceInvocationContext context = result.invocationContext();
+            context.attr(METRICS).set(new MetricsData(getRequestSize(result.content()), startTime));
+        } else {
+            metricConsumer.invocationComplete(
+                    result.encodedScheme().orElse(Scheme.of(SerializationFormat.UNKNOWN, sessionProtocol)),
+                    HttpResponseStatus.BAD_REQUEST.code(),
+                    lazyElapsedTime.getAsLong(), 0, 0,
```

Comment:
no other code uses `lazyElapsedTime`, it seems better to replace with `System.nanoTime() - startTime` directly.